File: 3211-generate-binary-strings-without-adjacent-zeros/3211-generate-binary-strings-without-adjacent-zeros.cpp
class Solution {
public:
    vector<string>ans;
    vector<string> validStrings(int n) {
        string temp = "";
        solve(n, temp);
        return ans;
    }
    void solve(int n, string temp) {
        if(n == 0) {
            ans.push_back(temp);
            return;
        }
        if(temp.empty()) {
            solve(n-1, temp + '0');
            solve(n-1, temp + '1');
        }
        else if(temp.back() == '1') {
            solve(n-1, temp + '1');
            solve(n-1, temp + '0');
        }
        else {
            solve(n-1, temp + '1');
        }
    }
};

File: 3211-generate-binary-strings-without-adjacent-zeros/3211-generate-binary-strings-without-adjacent-zeros.java
class Solution {
    List<String>ans = new ArrayList<>();
    public List<String> validStrings(int n) {
        StringBuilder temp = new StringBuilder();
        solve(n, temp);
        return ans;
    }
    void solve(int n, StringBuilder sb) {
        if(n == 0) {
            ans.add(sb.toString());
            return;
        }
        if(sb.isEmpty()) {
            solve(n-1, new StringBuilder(sb).append('0'));
            solve(n-1, new StringBuilder(sb).append('1'));
        }
        else if(sb.charAt(sb.length()-1) == '1') {
            solve(n-1, new StringBuilder(sb).append('0'));
            solve(n-1, new StringBuilder(sb).append('1'));
        }
        else {
            solve(n-1, new StringBuilder(sb).append('1'));
        }
    }
}